### PR TITLE
Add support for MSC3202: sending one-time key counts and fallback key usage states to Application Services.

### PR DESCRIPTION
**Pull request by reivilibre**
It's difficult to divide this one up any further (apart from minor things ~ 10 lines for config, etc, or leaving things unimplemented as stubs); if you can see a way that would do it effectively then let me know and I'll try to do that.

The general idea here is:
- Application Services will now receive one-time key counts and fallback key usage states for their users' devices
- The idea we settled on in #10653 is as follows:
    - Compute the set of 'interesting users' to the application service for this transaction, hereby defined as:
        - the AS's users that are in rooms which receive an event in this transaction; and
        - the Application Service `sender_localpart` (unsure if this has a more proper name, but it's vaguely the 'main user ID' for the application service and is important enough that it should not miss OTK/FBK changes)
     - Retrieve counts / FBK states for all of those and send them out

As discussed in #10653, doing it this way is easier implementation-wise (since otherwise we would need to somehow track changes) and is chattier than needed.

This should largely be a commit-by-commit review, though there are a few 'boring' commits.

[MSC3202 is available here.](https://github.com/matrix-org/matrix-doc/pull/3202)

### CODE REVIEW

File: synapse/appservice/scheduler.py
Comment (reivilibre) [reply synapse/appservice/scheduler.py:225]: Yeah, misleading word.

File: synapse/config/experimental.py
Comment (reivilibre) [reply synapse/config/experimental.py:67]: Happy to merge if you think that's best — I wasn't right sure because they're really quite separate features that could easily have been separate MSCs, but not fussed either way.

File: synapse/storage/databases/main/appservice.py
Comment (reivilibre) [reply synapse/storage/databases/main/appservice.py:363]: Bad comment. Plan for now is to leave this to a later piece of work focussed on 'reliability' (which is somewhat more important for to-device messages than OTKs/UFBKs as you ever only care about the latest state for the latter .

Have changed comment to more accurately reflect the plan.

File: synapse/appservice/scheduler.py
Comment (erikjohnston) [inline synapse/appservice/scheduler.py:252]: Can we move this into `_compute_msc3202_otk_counts_and_fallback_keys`? Rather than having two functions, one of which has a horrifically long function name :)

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (erikjohnston) [inline synapse/storage/databases/main/end_to_end_keys.py:481]:
```suggestion
                # algorithm will be None if this device has no keys.
                if algorithm is not None:
                    device_count_by_algo = result.setdefault(user_id, {}).setdefault(
                        device_id, {}
                    )
                    device_count_by_algo[algorithm] = count
```

Rather than creating empty dictionaries?

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (erikjohnston) [inline synapse/storage/databases/main/end_to_end_keys.py:532]: Ditto to above here

**Review by erikjohnston — COMMENTED** [review]
Otherwise LGTM

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (reivilibre) [reply synapse/storage/databases/main/end_to_end_keys.py:481]: I need to page in the MSC again (so good on you for mentioning, this could definitely use a comment), but I think we *want* the empty dicts to signal 'no keys'.

Comment (reivilibre) [reply synapse/storage/databases/main/end_to_end_keys.py:481]: Actually: the MSC says these can be omitted, but I think it's wrong. [I've opened a question to double check](https://github.com/matrix-org/matrix-doc/pull/3202/files#r807866130).

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (reivilibre) [reply synapse/storage/databases/main/end_to_end_keys.py:546]: I believe this is the same situation as unused fallback keys; if you run out of OTKs for a particular algorithm then you need to hear about it.

**Review by erikjohnston — APPROVED** [review]

File: synapse/appservice/api.py
Comment (Half-Shot) [inline synapse/appservice/api.py:278]: Having tried to play with this just now, I think this is wrong. The MSC states [`device_unused_fallback_key_types`](https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/otk-dl-appservice/proposals/3202-encrypted-appservices.md#proposal)

**Pull request merged**
